File: exercism/clojure/perfect-numbers/src/perfect_numbers.clj
(ns perfect-numbers)
(require '[clojure.string :as str])

(defn factors [n] 
  (loop [ factors [] current-factor 1 ]
    (if (= n current-factor)
      factors
      (recur (if (= 0 (mod n current-factor)) (cons current-factor factors) factors) (inc current-factor))
      )
    )
  )

(defn aliquot-sum [n]
  (reduce + (factors n))
  )

(defn classify [n] ;; <- arglist goes here
  (cond
    (<= n 0) (throw (IllegalArgumentException. "The provided number should be strictly positive"))
    (= n (aliquot-sum n)) :perfect
    (< n (aliquot-sum n)) :abundant
    (> n (aliquot-sum n)) :deficient
    )
  )

File: exercism/python/perfect-numbers/perfect_numbers.py
def classify(number):
    if (number <= 0):
        raise ValueError("Numbers less or equal to zero are not allowed")
    elif is_abundant(number):
        return "abundant"
    elif is_deficient(number):
        return "deficient"
    elif is_perfect(number):
        return "perfect"
    else:
       raise AssertionError("Value not allowed")

def is_perfect(number):
    return compute_divisors_sum(number) == number

def is_abundant(number):
    return compute_divisors_sum(number) > number

def is_deficient(number):
    return compute_divisors_sum(number) < number

def compute_divisors_sum(number):
    return sum([i for i in range(1,number) if number % i == 0])